src/wasmtime/component.rs

```diff
@@ -205,40 +272,166 @@ impl WasmValue for component::Val {
-        unwrap_val!(self, Self::String, "string").as_ref().into()
+        unwrap_val!(self, Self::String, "string").into()
     }
     fn unwrap_list(&self) -> Box<dyn Iterator<Item = Cow<Self>> + '_> {
         let list = unwrap_val!(self, Self::List, "list");
         Box::new(list.iter().map(cow))
     }
     fn unwrap_record(&self) -> Box<dyn Iterator<Item = (Cow<str>, Cow<Self>)> + '_> {
         let record = unwrap_val!(self, Self::Record, "record");
-        Box::new(record.fields().map(|(name, val)| (name.into(), cow(val))))
+        Box::new(record.iter().map(|(name, val)| (name.into(), cow(val))))
     }
     fn unwrap_tuple(&self) -> Box<dyn Iterator<Item = Cow<Self>> + '_> {
         let tuple = unwrap_val!(self, Self::Tuple, "tuple");
-        Box::new(tuple.values().iter().map(cow))
+        Box::new(tuple.iter().map(cow))
     }
     fn unwrap_variant(&self) -> (Cow<str>, Option<Cow<Self>>) {
-        let variant = unwrap_val!(self, Self::Variant, "variant");
-        (variant.discriminant().into(), variant.payload().map(cow))
+        let (discriminant, payload) = unwrap_2val!(self, Self::Variant, "variant");
+        (discriminant.into(), payload.as_deref().map(cow))
     }
     fn unwrap_enum(&self) -> Cow<str> {
-        unwrap_val!(self, Self::Enum, "enum").discriminant().into()
+        unwrap_val!(self, Self::Enum, "enum").into()
     }
     fn unwrap_option(&self) -> Option<Cow<Self>> {
-        unwrap_val!(self, Self::Option, "option").value().map(cow)
+        unwrap_val!(self, Self::Option, "option")
+            .as_deref()
+            .map(cow)
     }
     fn unwrap_result(&self) -> Result<Option<Cow<Self>>, Option<Cow<Self>>> {
-        match unwrap_val!(self, Self::Result, "result").value() {
-            Ok(val) => Ok(val.map(cow)),
-            Err(val) => Err(val.map(cow)),
+        match unwrap_val!(self, Self::Result, "result") {
+            Ok(t) => Ok(t.as_deref().map(cow)),
+            Err(e) => Err(e.as_deref().map(cow)),
         }
     }
     fn unwrap_flags(&self) -> Box<dyn Iterator<Item = Cow<str>> + '_> {
         let flags = unwrap_val!(self, Self::Flags, "flags");
-        Box::new(flags.flags().map(Into::into))
+        Box::new(flags.iter().map(Into::into))
+    }
+}
+
+// Returns an error if the given component::Val is not of the given component::Type.
+//
+// The component::Val::Resource(_) variant results in an unsupported error at this time.
+fn ensure_type_val(ty: &component::Type, val: &component::Val) -> Result<(), WasmValueError> {
+    let wrong_value_type =
+        || -> Result<(), WasmValueError> { Err(WasmValueError::wrong_value_type(ty, val)) };
+
+    if ty.kind() != val.kind() {
+        return wrong_value_type();
+    }
+
+    match val {
+        component::Val::List(vals) => {
+            let component::Type::List(list_handle) = ty else {
+                unreachable!(); // kind to kind check above should make this unreachable
+            };
```

Comment:
Wasmtime conveniently provides [helpers](https://docs.rs/wasmtime/20.0.2/wasmtime/component/enum.Type.html#method.unwrap_list) for these assertions:

```suggestion
            let list_handle = ty.unwrap_list();
```